plugin/trino-clickhouse/src/main/java/io/trino/plugin/clickhouse/ClickHouseClient.java

```diff
@@ -0,0 +1,389 @@
+                columnTypes.add(column.getColumnType());
+                jdbcColumnTypes.add(column.getJdbcTypeHandle());
+            }
+
+            copyTableSchema(connection, catalog, remoteSchema, remoteTable, tableName, columnNames.build());
+
+            return new JdbcOutputTableHandle(
+                    null,
+                    remoteSchema,
+                    remoteTable,
+                    columnNames.build(),
+                    columnTypes.build(),
+                    Optional.of(jdbcColumnTypes.build()),
+                    tableName);
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    protected String getColumnDefinitionSql(ConnectorSession session, ColumnMetadata column, String columnName)
+    {
+        StringBuilder sb = new StringBuilder()
+                .append(quoted(columnName))
+                .append(" ");
+        if (column.isNullable()) {
+            // set column nullable property explicitly
+            sb.append("Nullable(").append(toWriteMapping(session, column.getType()).getDataType()).append(")");
+        }
+        else {
+            // By default, the clickhouse column is not allowed to be null
+            sb.append(toWriteMapping(session, column.getType()).getDataType());
+        }
+        return sb.toString();
+    }
+
+    @Override
+    public void createSchema(ConnectorSession session, String schemaName)
+    {
+        execute(session, "CREATE DATABASE " + quoted(schemaName));
+    }
+
+    @Override
+    public void dropSchema(ConnectorSession session, String schemaName)
+    {
+        execute(session, "DROP DATABASE " + quoted(schemaName));
+    }
+
+    @Override
+    public void renameColumn(ConnectorSession session, JdbcTableHandle handle, JdbcColumnHandle jdbcColumn, String newColumnName)
+    {
+        try (Connection connection = connectionFactory.openConnection(session)) {
+            DatabaseMetaData metadata = connection.getMetaData();
+            if (metadata.storesUpperCaseIdentifiers()) {
+                newColumnName = newColumnName.toUpperCase(ENGLISH);
+            }
+            String sql = format("ALTER TABLE %s RENAME COLUMN %s TO %s ",
+                    handle.getRemoteTableName(),
+                    jdbcColumn.getColumnName(),
+                    newColumnName);
+            execute(connection, sql);
+        }
+        catch (SQLException e) {
+            throw new TrinoException(JDBC_ERROR, e);
+        }
+    }
+
+    @Override
+    protected ResultSet getTables(Connection connection, Optional<String> schemaName, Optional<String> tableName)
+            throws SQLException
+    {
+        // ClickHouse maps their "database" to SQL catalogs and does not have schemas
+        DatabaseMetaData metadata = connection.getMetaData();
+        return metadata.getTables(
+                null,
+                schemaName.orElse(null),
```

Comment:
- per the comment, `schemaName` would go as first parameter here (the `catalog`), and not the second `schemaPattern`). Is the comment wrong, or the code?
- `schemaPattern` is a pattern, so should be `escapeNamePattern`